src/utils/index.test.tsx

```diff
@@ -0,0 +1,279 @@
+import { InstrumentName } from '@/features/synth/types'
+import {
+  breakpoints,
+  clamp,
+  convertHexColorToIntArr,
+  Deferred,
+  diffKeys,
+  fileToString,
+  fileToUint8,
+  formatInstrumentName,
+  formatTime,
+  getHands,
+  isFileMidi,
+  isFileXML,
+  isMobile,
+  isNumber,
+  mapValues,
+  pickHex,
+  range,
+  round,
+  sleep,
+} from './index'
+
+describe('isMobile', () => {
+  afterEach(() => {
+    jest.resetAllMocks()
+  })
+
+  it('should return false if not in a mobile environment based on userAgent', () => {
+    const userAgent =
+      'Mozilla/5.0 (Macintosh; Intel Mac OS X 10_15_7) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/116.0.0.0 Safari/537.36'
+    Object.defineProperty(window.navigator, 'userAgent', { value: userAgent, configurable: true })
+    expect(isMobile()).toBe(false)
+  })
+
+  it.each(['iPhone', 'iPad', 'iPod', 'Android'])(
+    'should return mobile as true for userAgent %s',
+    (userAgent) => {
+      Object.defineProperty(window.navigator, 'userAgent', { value: userAgent, configurable: true })
+      expect(isMobile()).toBe(true)
+    },
+  )
+})
+
+describe('sleep', () => {
+  beforeEach(() => {
+    jest.useFakeTimers()
+  })
+
+  afterEach(() => {
+    jest.useRealTimers()
+  })
+
+  it('should resolve after the specified time', async () => {
+    const MS_1000 = 1000
+
+    const callback = jest.fn()
+    sleep(MS_1000).then(callback)
+
+    expect(callback).not.toBeCalled()
+
+    // only after MS_1000
+    jest.advanceTimersByTime(MS_1000)
+
+    // This allows pending promises to resolve
+    await Promise.resolve()
+    expect(callback).toBeCalled()
+  })
+})
+
+describe('range', () => {
+  it('should return an array of numbers from start to end', () => {
+    const start = 1
+    const end = 5
+    const expected = [1, 2, 3, 4, 5]
+    expect(range(start, end)).toEqual(expected)
+  })
+
+  it('should return an empty array if start is greater than end', () => {
+    const start = 5
+    const end = 1
+    expect(range(start, end)).toEqual([])
+  })
+
+  it('should return an array with a single element if start equals end', () => {
+    const start = 5
+    const end = 5
+    expect(range(start, end)).toEqual([5])
+  })
+})
+
+describe('formatTime', () => {
+  it('should throw an error if seconds is a string or undefined', () => {
+    expect(() => formatTime('100')).toThrow('Should not call formatTime on a string')
+    expect(() => formatTime(undefined)).toThrow('Should not call formatTime on a string')
```

Comment:
This looks like a bug. It was called on `undefined` not a string!

Reply:
This was testing that error message, should I change the message for undefined?

```
 if (typeof seconds === 'string' || seconds === undefined) {
    throw new Error('Should not call formatTime on a string')
  }
  ```

Reply:
What if we just use Type Script for that check? I think it would handle as we expect, right?

So we wouldn't need this check, neither this test.

Reply:
Wow! Thanks to TypeScript, we might have found a bug?

```
./src/features/pages/SelectSong/index.tsx:107:43
Type error: Argument of type 'string | number' is not assignable to parameter of type 'number'.
  Type 'string' is not assignable to type 'number'.

  105 |                 label: 'Length',
  106 |                 id: 'duration',
> 107 |                 format: (n) => formatTime(n),
      |                                           ^
  108 |               },
  109 |               { label: 'Source', id: 'source' },
  110 |             ]}
  ```

Reply:
I don't know what this table is formatting here. I'll check on later.
I would either cast to number, or check type if string, to not format.

```
format: (n) => formatTime(Number(n)),
```


or 
```
format: (n) => isNumber(n) ? formatTime(n) : n,
```